Instructions for live documentation with KLIPSE

viebel opened the pull request:
Codox really rocks.

I've been able to integrate [KLIPSE](https://github.com/viebel/klipse) for  live documentation on my [gadjett library] (https://github.com/viebel/gadjett) in less than an hour.

Take a look at the result here: http://viebel.github.io/gadjett/gadjett.collections.html#var-sequence-.3Emap

(It takes a couple of seconds to load the code at run-time!)

Codox is simple and highly configurable making it really smooth to tweak.

Please review the instructions that I've written.

I believe there should be a simpler way to integrate Klipse with Codox - especially around the `:require` statement that is currently created manually with the names of all the functions of the namespace (in `ClojureScript`, there is no way to refer all the namespace public vars).

weavejester commented:
Nice work. What does the `~~~` denote? Is it just another way of starting a code block?

Rather than use `gadjett` as the example namespace and project, let's change it to `example`.

It would also be nice if you reduced the horizontal size of the example code you use, so it can fit on a page.

viebel commented:
1. `~~~klipse` starts a code block with `class="klipse"`
2. I have added` lineWrapping: true` in the `klipse_settings` to fit the code on a page
3. I don't know how to change `gadjett` to `example`: what about the functions?

xsc commented:
Just throwing this in here: This might be a good candidate for the (hopefully) upcoming theme support (discussion in #135) since it's just a set of predefined transformations.

Wouldn't it be nicer if – instead of having to add these transformations by hand – the user would just include a dependency and set (for example) `:codox {:theme [:default :klipse]}`?

weavejester commented:
> I have `addedlineWrapping: true` in the `klipse_settings` to fit the code on a page

I meant line width in the README. Try to keep within 72 characters for text, a little longer if necessary for code.

> I don't know how to change `gadjett` to `example`: what about the functions?

Maybe I've misunderstood, but I thought `gadjett` was just used as an example? Is there something in that library that people need to get Klipse working?

viebel commented:
I have updated the `README` with `my_user`, `my_repo`, `my_ns` and `my_func` and it fits into a line width.

Theme support will be awesome.

viebel commented:
Do you want to merge the Pull Request?

weavejester commented:
I was waiting for you to implement the previous changes. The line width in the README is longer than 72 characters, for instance.

I also wonder whether or not this would be better done as a theme, rather than manual instructions. Once the theme PR is fully tested and merged, we should be able to write a Klipse theme to mix in.

viebel commented:
The themes are really nice!

I have added the klipse theme in codox and updated the instructions accordingly.


It could be even nicer, if it was possible from `theme.edn` to access values from the `project.clj`.


I hope that the line width is ok now.

weavejester commented:
> I have added the klipse theme in codox and updated the instructions accordingly.

Thanks! I'll take a look.

> It could be even nicer, if it was possible from `theme.edn` to access values from the `project.clj`.

That's an interesting idea. What's your use-case?

weavejester commented on boot-codox/src/codox/boot.clj:
```diff
@@ -6,7 +6,7 @@
 
 (defn- pod-deps []
-  (remove pod/dependency-loaded? '[[codox "0.10.0"]]))
+  (remove pod/dependency-loaded? '[[codox "0.11.0"]]))
```
This change needs to be reverted. PRs should only contain code related to the feature they implement.

weavejester commented on codox/project.clj:
```diff
@@ -1,4 +1,4 @@
-(defproject codox "0.10.0"
+(defproject codox "0.11.0"
```
This change also needs to be reverted.

weavejester commented on codox/src/codox/reader/clojurescript.clj:
```diff
@@ -6,9 +6,14 @@
             [cljs.analyzer.api :as ana]
             [clojure.string :as str]))
 
+(defn- cljs-filename? [filename]
```
This change seems unrelated to the Klipse theme, and therefore needs to be reverted.

viebel replied:
Do you want me to open another pull request for supporting both `cljc` and `cljs` files in clojurescript mode?

weavejester commented on lein-codox/project.clj:
```diff
@@ -1,8 +1,13 @@
-(defproject lein-codox "0.10.0"
-  :description "Codox Leiningen plugin"
+(defproject codox "0.11.0"
```
This looks like a mistake. You've overwritten the lein-codox project file with the codox project file for some reason.

weavejester commented on lein-codox/src/leiningen/codox.clj:
```diff
@@ -24,7 +24,7 @@
         options (get-options project)]
     (eval/eval-in-project
-     (deps/add-if-missing project '[codox "0.10.0"])
+     (deps/add-if-missing project '[codox "0.11.0"])
```
This change needs to be reverted.

weavejester commented on codox/resources/codox/theme/klipse/theme.edn:
```diff
@@ -0,0 +1,19 @@
+{:transforms
+                 [[:body]
```
Shouldn't this be appending to the `:head`?

viebel replied:
No. KLIPSE must be loaded at the end of the `body`.

weavejester replied:
Even the stylesheet?

viebel commented:
Take a look at the [project.clj of gadjett](https://github.com/viebel/gadjett/blob/master/project.clj): I still have to do a couple of html transformations.

Those transformations are required in order to prepend an element with clojurescript code that requires the namespaces to test.

It could be templated with two parameters:

- `:github-repo` for `:data-external-libs`
- `:require-statement` for  the content of the `div`

It would be nice to read those parameters from `project.clj` instead of requiring the user to write all the boilerplate code.

Makes sense?

weavejester commented:
> It would be nice to read those parameters from `project.clj` instead of requiring the user to write all the boilerplate code.

I see now. That makes sense. Let me think about the best way to do that :)

viebel commented:
OK. Anyway, after the review, could you merge the PR?

weavejester commented:
I think right now it feels like we're not quite there. Themes give us a way of removing some boilerplate, but not all. If themes supported parameters of some description, we could remove all boilerplate, and that seems like a worthwhile goal to pursue.

For instance, we might have something like:

```clojure
:themes [:default
         [:klipse {:external-lib "https://raw.githubusercontent.com/my_user/my_repo/master/src"
                   :require      "[my-repo.core :refer :all]"}]]
```

I'd like to put this PR on pause briefly while I investigate the best way to extend themes with parameters. I don't want to get into a situation where we add information to the README, only for it to be made obsolete a few days later.

viebel commented:
Sure. 

Let me know, when you found the  best way to extend themes with parameters.

Meanwhile, I will address the reviews you requested.

weavejester commented:
Thanks for being patient. 👍

viebel commented:
PR updated - addressed all the review remarks.
Did you find  a good way to extend themes with parameters?

viebel commented:
@weavejester any update on this PR?

weavejester commented:
I've added parameter support for themes. Can you try it out and see if it fits your needs?

viebel commented:
Yeah the way you did it works fine. But in order to use it in  my `gadjett` lib, I need you to merge #141

weavejester commented:
Does the current master branch of Codox work with your Klipse theme and project, now?

viebel commented:
Yes. It works like a charm 👍 
Can you merge it?

weavejester commented:
I was thinking about whether it would be better including in Codox by default, or whether it would make more sense as a library that could optionally be included. I'm currently leaning toward having it as a separate library, and then I could link it from the Codox README (e.g. "If you want Klipse support, use this third-party library"). What do you think?

viebel commented:
You are talking about the theme file for klipse?
For me, it would have much more impact if you include it in the Codox
README.

The fact that it is in Codox README is kind of a vote of confidence for
KLIPSE.

So I'd really appreciate have KLIPSE inlined in Codox README.

But obviously, you decide...

On Sat, 8 Oct 2016 at 20:46 James Reeves <notifications@github.com> wrote:

> I was thinking about whether it would be better including in Codox by
> default, or whether it would make more sense as a library that could
> optionally be included. I'm currently leaning toward having it as a
> separate library, and then I could link it from the Codox README (e.g. "If
> you want Klipse support, use this third-party library"). What do you think?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/weavejester/codox/pull/137#issuecomment-252438358>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6VPrWD6hSHYp1YnY3o4apusu2X1uCfks5qx9cSgaJpZM4J0eTT>
> .
>

weavejester commented:
I'd still put it in the core README, but as a link. Something like:

```markdown
## Live Documentation

You can make the code in your documentation live and interactive by
using the [Klipse][] theme from [Yehonathan Sharvit][].
```

Then in the theme README you could have a more detailed explanation. How about that?

viebel commented:
And how people are going to have the klipse theme file?
Will they have to download it into their repo?

On Sat, 8 Oct 2016 at 21:01 James Reeves <notifications@github.com> wrote:

> I'd still put it in the core README, but as a link. Something like:
>
> ## Live Documentation
>
> You can make the code in your documentation live and interactive by
> using the [Klipse][] theme from [Yehonathan Sharvit][].
>
> Then in the theme README you could have a more detailed explanation. How
> about that?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/weavejester/codox/pull/137#issuecomment-252439086>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6VPs0Ne0Bv4XUHXVq8oonQg7ifOG39ks5qx9psgaJpZM4J0eTT>
> .
>

weavejester commented:
Just release it as its own library. So create a new project (`codox-klipse-theme`?), put the theme file in the right `resources` subdirectory, then deploy it to Clojars. People can then add it as a dependency to use it.

viebel commented:
I see.
Indeed it makes more sense to have it in a separate lib.
Then I can update new version without asking you to merge my PRs.

Will create one now

On Sat, 8 Oct 2016 at 21:14 James Reeves <notifications@github.com> wrote:

> Just release it as its own library. So create a new project (
> codox-klipse-theme?), put the theme file in the right resources
> subdirectory, then deploy it to Clojars. People can then add it as a
> dependency to use it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/weavejester/codox/pull/137#issuecomment-252439705>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6VPvvbWP3mQC5hPG79g5jez7T2PKUnks5qx91rgaJpZM4J0eTT>
> .
>

viebel commented:
Here it is https://clojars.org/viebel/codox-klipse-theme

So I can close my PR.

Let me know viebel@gmail.com when  you to add a note about the klipse theme in the README.

weavejester commented:
I'll add a link to the README. Do you think you should add something to your README about how to install it? Even if it's just as simple as: add this line to your dev dependencies.

viebel commented:
Thanks you. I updated the README.

On Sat, 8 Oct 2016 at 21:43 James Reeves <notifications@github.com> wrote:

> I'll add a link to the README. Do you think you should add something to
> your README about how to install it? Even if it's just as simple as: add
> this line to your dev dependencies.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/weavejester/codox/pull/137#issuecomment-252441194>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6VPiIMvbXjZVMZSIAObVFbF4Rwuak_ks5qx-RagaJpZM4J0eTT>
> .
>

weavejester commented:
@viebel, I added the README link.

viebel commented:
Thx. Do you have a library of your own where you could use klipse codox
theme?

On Sun, Oct 9, 2016, 17:42 James Reeves <notifications@github.com> wrote:

> @viebel <https://github.com/viebel>, I added the README link.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/weavejester/codox/pull/137#issuecomment-252490338>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6VPn8vOmmQKyHta1A30LArh9K5iTnXks5qyP1MgaJpZM4J0eTT>
> .
>
